lib/qmi.c

```diff
@@ -794,20 +779,44 @@ ssize_t qmi_encode_message(struct qrtr_packet *pkt, int type, int msg_id,
 	return pkt->data_len;
 }
 
-int qmi_decode_header(const struct qrtr_packet *pkt, unsigned int *msg_id)
+const struct qmi_header *qmi_get_header(const struct qrtr_packet *pkt)
 {
 	const struct qmi_header *qmi = pkt->data;
 
 	if (qmi->msg_len != pkt->data_len - sizeof(*qmi)) {
-		LOGW("[RMTFS] Invalid length of incoming qmi request\n");
-		return -EINVAL;
+		LOGW("[QRTR] Invalid length of incoming qmi request\n");
+		return NULL;
 	}
 
+	return qmi;
+}
+
+int qmi_decode_header(const struct qrtr_packet *pkt, unsigned int *msg_id)
+{
+	const struct qmi_header *qmi = qmi_get_header(pkt);
+
 	*msg_id = qmi->msg_id;
 
 	return 0;
 }
 
+int qmi_decode_header2(const struct qrtr_packet *pkt, unsigned int *msg_id, unsigned char *type,
+		       unsigned short *txn_id)
+{
+	const struct qmi_header *qmi = qmi_get_header(pkt);
+	if (!qmi)
+		return -1;
```

Comment:
Currently these errors return -EINVAL, it could be argued that returning -1 and setting errno is more idiomatic - but I think you should stick to the current semantics.